pyteal/ast/subroutine.py

```diff
@@ -720,24 +725,21 @@ def var_n_loaded(
         raise TealInputError(
             f"Subroutine function does not return a PyTeal expression. Got type {type(subroutine_body)}."
         )
+
+    deferred_expr: Optional[Expr] = None
+
     # if there is an output keyword argument for ABI, place the storing on the stack
     if output_carrying_abi:
-        if subroutine_body.has_return():
-            raise TealInputError(
```

Comment:
Are there any downstream implications for removing this check? I removed the check in my own branch and saw that no unit tests failed. This either indicates that this was a dead code, or that we didn't have coverage.

Reply:
Here's the story, referring back to this discussion: https://github.com/algorand/pyteal/pull/256/files#r865461530, where Jason pointed out that `has_return` is true *only* when every possible codepath in expr results in a return.

The code above thus does not quite work, with an example:
supposing the codeblock has an `If` expr with `then` branch has return, and `else` branch no return, then `has_return` is not returning true.

- So is this a deadcode, I don't think so.
- Is it wrong, yea it is a misuse of `has_return`.
- Is there coverage not enough, I think that's also right, but the restriction of no `return` expr in subroutine codeblock is lifted, so the coverage does not matter here.

What we need to do in abi-subroutine PR is to introduce subroutine with conditional return, as Jason suggested.

Reply:
Resolving because this should be addressed in #256